tests/system_tests/wandb_backend_spy/proxy.py

```diff
@@ -0,0 +1,212 @@
+from __future__ import annotations
+
+import asyncio
+import contextlib
+import threading
+import time
+import traceback
+from typing import Iterator
+
+import fastapi
+import httpx
+import uvicorn
+
+from .spy import WandbBackendSpy
+
+
+@contextlib.contextmanager
+def spy_proxy(
+    proxy_port: int,
+    target_host: str,
+    target_port: int,
+) -> Iterator[WandbBackendProxy]:
+    """A context manager that proxies a W&B backend.
+
+    The proxy server will be available at 127.0.0.1 on the given port.
+    The server is shut down when the context manager exits.
+
+    Args:
+        proxy_port: The port on which to run the proxy server.
+        target_host: The hostname of the W&B backend to proxy.
+        target_port: The port of the W&B backend to proxy.
+    """
+
+    http_client = httpx.AsyncClient()
+    spy = WandbBackendProxy(
+        client=http_client,
+        target_host=target_host,
+        target_port=target_port,
+    )
+    server = uvicorn.Server(uvicorn.Config(spy._to_fast_api(), port=proxy_port))
+
+    proxy_thread = threading.Thread(
+        target=asyncio.run,
+        args=[_serve_then_close(server, http_client)],
+    )
+    proxy_thread.start()
+    _wait_for_server(server)
+
+    try:
+        yield spy
+    finally:
+        try:
+            server.should_exit = True
+            proxy_thread.join(timeout=30)
+        except TimeoutError as e:
+            raise AssertionError("Backend proxy server failed to shut down.") from e
+
+
+async def _serve_then_close(
+    server: uvicorn.Server,
+    http_client: httpx.AsyncClient,
+) -> None:
+    """Start the server, wait for it to shut down, then clean up."""
+    await server.serve()
+    await http_client.aclose()
+
+
+def _wait_for_server(server: uvicorn.Server) -> None:
+    """Wait until the server has started.
+
+    Raises:
+        AssertionError: if the server does not start within a timeout.
+    """
+    tries = 100
+
+    while tries > 0 and not server.started:
+        time.sleep(0.1)
+        tries -= 1
```

Comment:
Any reason why we didn't just throw an exception after some amount of time has elapsed?
I feel like its a little clearer and easier if we ever want to modify the timeout, to know exactly how much possibly time we are adding

e.g.
```
from datetime import datetime, timedelta

now = datetime.now()
end_time = now + timedelta(seconds=timeout)

while not datetime.now() > end_time and not server.started:
    time.sleep(0.1)
```

Reply:
you're right, that is clearer! will update

Reply:
done; I use `time.monotonic()` rather than `datetime` since that's more appropriate for this kind of use case